### PR TITLE
BUG: maybe_convert_objects mixed datetimes and timedeltas 

### PR DESCRIPTION
**Pull request by BeforeFlight**
- [X] corresponding [discussion](https://github.com/pandas-dev/pandas/issues/27417#issuecomment-512014358) 27417
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

In order to stick to `maybe_convert_objects` realization it was needed to introduce `Seen().nat_` param. Since we want to distinguish following situations: were seen `NaT` or `datetime` or `timedelta`

NOTE: in current realization array full of `NaT` will be converted to `np.datetime64`.

### CODE REVIEW

**Comment by BeforeFlight** [comment]
Have also ran following comparing with old realization tests:
```python
check_types = [None, np.nan, pd.NaT,
               True,
               np.iinfo(np.int64).min-1, -1, 1, np.iinfo(np.int64).max+1,
               np.iinfo(np.uint64).max, np.iinfo(np.uint64).max+1,
               1.1, np.float_(1.1),
               1j, np.complex_(1j),
               datetime(2000, 1, 1),
               datetime(2000, 1, 1, 0, 0, 0, 0, timezone.utc),
               np.datetime64('2000-01-01'),
               timedelta(365),
               np.timedelta64(1, 's'),
               # Decimal(1.1),
               'object'
               ]
DT_TD_NaT_inds = [2] + list(range(14, 19))


def data_gen(perms, exclude_all_is_time=False):
    data = np.array([0]*N, dtype='object')
    for perm in perms:
        if exclude_all_is_time and all(ind in DT_TD_NaT_inds for ind in perm):
            continue
        for i in range(data.shape[0]):
            data[i] = check_types[perm[i]]
        yield data.copy()


N = 3
exclude_all_is_time = False
perms = itertools.product(list(range(len(check_types))), repeat=N)
datas = data_gen(perms, exclude_all_is_time)


class Tests:
    @pytest.mark.parametrize("data", datas)
    @pytest.mark.parametrize("safe", [1])
    # @pytest.mark.parametrize("conv_DT", [1])
    # @pytest.mark.parametrize("conv_TD", [1])
    # @pytest.mark.parametrize("safe", [0, 1])
    @pytest.mark.parametrize("conv_DT", [0, 1])
    @pytest.mark.parametrize("conv_TD", [0, 1])
    def test_maybe_convert_objects(self, data, safe, conv_DT, conv_TD):
        old = lib.maybe_convert_objects_OLD(data, 1, safe, conv_DT, conv_TD)
        new = lib.maybe_convert_objects(data, 1, safe, conv_DT, conv_TD)
        print('data:', data, data.dtype,
              '\nold:', old, old.dtype,
              '\nnew:', new, new.dtype)
        if isinstance(old, np.ndarray):
            tm.assert_numpy_array_equal(old, new)
        else:
            tm.assert_index_equal(old, new)
```
NOTE: have not parametrized by `safe` flag cause it concerns only `numeric`.

While running this tests with `exclude_all_is_time`, i.e. excluding tests with generated array fully consisting of `NaT` or `datetime` or `timedelta`- there is no errors all tests passed.

While `exclude_all_is_time` disabled (full test suite) got
```
Results (82.92s):
    31910 passed
      90 failed
```
Visually inspecting all of failed all of them concerns only BUG that is being fixed.

**Comment by BeforeFlight** [comment]
Also have a question that is in general non relevant but don't understand where to ask. While adding `Seen.nat_` I've changed following functions as well:
https://github.com/pandas-dev/pandas/blob/2ae5152690cea598a7f377eb4cb264d6d42bb0cc/pandas/_libs/lib.pyx#L1045-L1051

My question is why in these functions there is no `.null_` nor `datetimetz_` nor `.object_`?

**Comment by jbrockmendel** [comment]
can you add an xref to the top post pointing back to the discussion (even though this doesnt close that issue)

**Comment by jbrockmendel** [comment]
needs tests.  even if they're not passing yet please push them

> My question is why in these functions there is no .null_ nor datetimetz_ nor .object_?

I don't know off the top of my head.  Probably check the blame to see where this was implemented?

**Comment by BeforeFlight** [comment]
> can you add an xref to the top post pointing back to the discussion (even though this doesnt close that issue)

Check please.

> needs tests. even if they're not passing yet please push them

But in order to test like I did i need both realization old and new in the `lib.pyx`. On my laptop I did tests -> remove old `maybe_convert_objects` -> push and make PR. How it should be tested then with old function removed?

**Comment by BeforeFlight** [comment]
The only solution (_strange_) for tests I see here is to dump tests output into file -> make function changes -> use same tests in order to comparing outputs. Then if changes will be approved this file with tests output should be updated to new outputs as well.

**Comment by gfyoung** [comment]
`My question is why in these functions there is no .null_ nor datetimetz_ nor .object_?`

@BeforeFlight : These were implemented initially to combat large number overflow issues.  Good call though to use them for purposes beyond that.

**Comment by jbrockmendel** [comment]
@BeforeFlight possibly helpful: coverage results for lib.pyx showing which lines are not hit in the tests:

pandas/_libs/lib.pyx                         1084     70    94%   307, 370, 651, 655, 658, 699-705, 707-710, 718-732, 828, 978, 983, 988, 1245-1246, 1268, 1445, 1519-1521, 1641, 1644-1646, 1696, 1718, 1889-1890, 1903, 1935, 1944, 1954-1956, 2006-2007, 2009-2010, 2022-2024, 2128-2131, 2135-2136, 2138-2139, 2143, 2145-2146, 2148-2153, 2191

grain of salt, since cython coverage results have been buggy in the past

**Comment by BeforeFlight** [comment]
@jbrockmendel so how should I add tests? Copy old function in tests section?

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2046,11 +2049,18 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
             else:
                 if not seen.bool_:
                     if seen.datetime_:
-                        if not seen.numeric_:
-                            return datetimes
+                        if not seen.timedelta_:
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: can you make this 1 level higher elif's, its very hard to follow these branches currently

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -2077,11 +2087,18 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
...
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: same as above

**Review by jreback — CHANGES_REQUESTED** [review]
this needs tests even before a change. its very hard to tell if what you did actually changed anything.

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (BeforeFlight) [reply pandas/_libs/lib.pyx]: @jreback I've proposed simple test in corresponding issue that returns undesired output - should I add this single test only? If not - `maybe_convert_objects` need full test suite (for _all_ possible types combinations) but as I tried to explain earlier  there will be thousands of it which is not doable 'by hands'.

Comment (jreback) [reply pandas/_libs/lib.pyx]: we already have quite a few tests
i am ok with systematic testing but don’t go overboard

so happy to merge tests first

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (BeforeFlight) [reply pandas/_libs/lib.pyx]: Will do. As for me - this nested "if"s are ugly even before I've changed anything. Besides it is one of the good places to make a mistake (as it happens in current case).

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (BeforeFlight) [reply pandas/_libs/lib.pyx]: > so happy to merge tests first

@jreback sorry but from your answer I still can't understand - is single test which is fixed by this PR enough or not?

Comment (jreback) [reply pandas/_libs/lib.pyx]: yes this needs at least a single test

if u want to do more comprehensive tests that xfail some (that re not passing now) is ok too (can be before or after this PR)

Comment (BeforeFlight) [reply pandas/_libs/lib.pyx]: Will add. 

The problem with comprehensive tests:  it cannot be done by hands -> it needs some "trusted" function to tested against i.e. needs copying old function for tests purposes only (as one of the options) - and here I don't understand may I do so or not.

**Comment by jbrockmendel** [comment]
> so how should I add tests? Copy old function in tests section?

There should be some managably-sized set of cases that you were trying to fix here, i.e. that were previous wrong but are now right.  Write those tests specifically.  No need to try have a zillion tests.

**Comment by BeforeFlight** [comment]
> No need to try have a zillion tests

I get it. Agree that for this PR bunch of tests should be enough. 

But at the same time I should mark that inconsistencies of `maybe_convert_objects` that I've found - all was result of thoroughly testing. 
And another point to mention - this is core function (well I guess). That is why such complete test suite is needed in my opinion (not zillion tests - but _complete_ tests; count shouldn't matter). Besides this test suite may be ran only when this concrete function is amended.

**Comment by jbrockmendel** [comment]
Improving the thoroughness of the test suite is great, but for a bugfix PR, the tests should be specific to the bug being fixed.

File: pandas/tests/dtypes/test_inference.py [hunk 3]
```diff
@@ -531,6 +531,15 @@ def test_maybe_convert_objects_uint64(self):
         exp = np.array([2 ** 63, -1], dtype=object)
         tm.assert_numpy_array_equal(lib.maybe_convert_objects(arr), exp)
 
+    def test_maybe_convert_objects_datetime(self):
+        # GH27438
+        arr = np.array([0, 0], dtype=object)
+        arr[0] = np.datetime64('2000-01-01')
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: this construction seems odd, you can just construct directly no? e.g.
```arr = np.array([np.datetime...., np.timedetla.lll], dtype='object'])```

Comment (jreback) [reply pandas/tests/dtypes/test_inference.py]: don't you need a NaT somewhere? what about throwing in a np.nan as well (maybe another case)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/dtypes/test_inference.py (hunk 3 above)
Comment (BeforeFlight) [reply pandas/tests/dtypes/test_inference.py]: Construction is odd, yes - will rewrite (it's leftover of previous tests sorry). Will add tests with NaT as well. 

As for np.nan - not quite understand. Add something with datetimes and np.nan?

Comment (jreback) [reply pandas/tests/dtypes/test_inference.py]: yes we likely already have a test for this, but pls check

Comment (BeforeFlight) [reply pandas/tests/dtypes/test_inference.py]: @jreback all is green please check.

Now there are 10 tests for `maybe_convert_objects` + 13 tests through `Dataframe` constructor (there are also tests for `maybe_convert_numeric` but I did not count these).

Comment (BeforeFlight) [reply pandas/tests/dtypes/test_inference.py]: Edit by the way - no one confirm my changes with array full of `NaT` here - now it returns `datetime` without any warning of ambiguity with `timedelta`. Check it pls too.

**Comment by BeforeFlight** [comment]
Should I just remove this `NaT`s ambiguity for now and return `object`?

**Comment by jbrockmendel** [comment]
> Should I just remove this NaTs ambiguity for now and return object? Cause this concrete PR is somewhat not relative to this NaTs ambiguity problem.

Between here and the issue you're bringing up a lot of questions and I'm having trouble keeping track of them.  To the extent that you can focus on one unambiguous bug at a time, that will make things easier for reviewers.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@BeforeFlight lgtm. does this have *any* user facing effects at all? meaning is there an incorrect Series/DataFrame construction currently?

**Comment by BeforeFlight** [comment]
Have to check (I'm not aware of all calling to `maybe_convert_object`) but if it used by some constructor directly - it might produce undesired results. The example that leads to this bug explains what I mean:
```python
import numpy as np
import pandas as pd
import pandas._libs.lib as lib

data = np.array([0]*2, dtype='object')
data[0] = np.datetime64('2000-01-01')
data[1] = np.timedelta64(1, 's')
old = lib.maybe_convert_objects(data,1,1,1,1)
print(old, old.dtype)
```
Output:
```python
['2000-01-01T00:00:00.000000000' '1970-01-01T00:00:00.021389104'] datetime64[ns]
```

As I already mention at issue branch - datetime array is actually not initialized 2nd value at all (left it as it is after `np.empty`) -> might be any other in theory. 

Will try to reproduce it on some constructor.

**Comment by BeforeFlight** [comment]
No, seems like there is no any effects for user at all in current code state. The reason here is that in order to produce such bug both flags `convert_datetime` and `convert_timedelta` must be `True`. But I've search and visually inspect all `lib.maybe_convert_objects` calls in `pandas` folder - there is no such case. Only separate flags `convert_datetime` or `convert_timedelta` but not both at same time.

**Comment by jreback** [comment]
thanks @BeforeFlight
